src/java/org/apache/cassandra/cql3/restrictions/MergedRestriction.java

```diff
@@ -0,0 +1,345 @@
+    }
+
+    @Override
+    public boolean isEQ() {
+        return false; // For the moment we do not support merging EQ restriction with anything else.
+    }
+
+    @Override
+    public boolean isIN()
+    {
+        return false; // For the moment we do not support merging IN restriction with anything else.
+    }
+
+    @Override
+    public boolean isANN() {
+        return false; // For the moment we do not support merging ANN restriction with anything else.
+    }
+
+    @Override
+    public boolean isSlice()
+    {
+        return isSlice;
+    }
+
+    @Override
+    public boolean isColumnLevel() {
+        return false;
+    }
+
+    @Override
+    public ColumnMetadata firstColumn()
+    {
+        return columns.get(0);
+    }
+
+    @Override
+    public ColumnMetadata lastColumn()
+    {
+        return columns.get(columns.size() - 1);
+    }
+
+    @Override
+    public List<ColumnMetadata> columns()
+    {
+        return columns;
+    }
+
+    @Override
+    public void addFunctionsTo(List<Function> functions)
+    {
+        for (int i = 0, m = restrictions.size(); i < m; i++)
+        {
+            restrictions.get(i).addFunctionsTo(functions);
+        }
+    }
+
+    @Override
+    public boolean needsFilteringOrIndexing()
+    {
+        for (int i = 0, m = restrictions.size(); i < m; i++)
+        {
+            if (restrictions.get(i).needsFilteringOrIndexing())
+                return true;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean needsFiltering(Index.Group indexGroup)
+    {
+        // multiple contains might require filtering on some indexes, since that is equivalent to a disjunction (or)
+        boolean hasMultipleContains = containsCount > 1;
+
+        for (Index index : indexGroup.getIndexes())
+        {
+            if (isSupportedBy(index) && !(hasMultipleContains && index.filtersMultipleContains()))
+                return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public Index findSupportingIndex(Iterable<Index> indexes)
+    {
+        for (int i = 0, m = restrictions.size(); i < m; i++)
+        {
+            Index index = restrictions.get(i).findSupportingIndex(indexes);
+            if (index != null)
+                return index;
+        }
+        return null;
+    }
+
+    @Override
+    public boolean isSupportedBy(Index index)
+    {
+        for (SingleRestriction restriction : restrictions)
+        {
+            if (restriction.isSupportedBy(index))
+                return true;
+        }
+        return false;
+    }
+
+    @Override
+    public List<ClusteringElements> values(QueryOptions options)
+    {
+        List<ClusteringElements> values = restrictions.get(0).values(options);
+        for (int i = 1, m = restrictions.size(); i < m; i++)
+        {
+            values.retainAll(restrictions.get(i).values(options));
+        }
+        return values;
```

Comment:
The returned values here might not be actual clusterings, right? I think the JavaDoc for `ClusteringElements` mentions this, but the naming just threw me off at first glance...

Reply:
I agree the naming is confusing. It is the consequence of the fact that we use `MultiCBuilder` and `Clustering` in the code to compute ByteBuffers for partition keys and token.

Reply:
I also did not find a better name which would not have made things more confusing at other places.